### PR TITLE
Detach resource group slot on segment

### PR DESCRIPTION
**Pull request by xiong-gang**
QE detach resource group slot at the end of transaction, the
    last QE of the slot release the slot and release the overused
    memory if resource group config has been changed.

### CODE REVIEW

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -2022,134 +2053,68 @@ UnassignResGroupOnMaster(void)
...
 	/* Init slot */
-	slot = self->slot;
-	slot->sessionId = gp_session_id;
-	slot->caps = caps;
-	slot->memQuota = slotGetMemQuotaExpected(&caps);
-	ResGroupSetMemorySpillRatio(&caps);
-	Assert(slot->memQuota > 0);
-
-	if (prevGroup != group || prevSlot != slot)
+	slot = &group->slots[self->slotId];
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: `selfSetSlot()` will set `self->slot` properly, so we can use `slot = self->slot` here.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -2022,134 +2053,68 @@ UnassignResGroupOnMaster(void)
...
-		groupAcquireMemQuota(group, &slot->caps);
-		LWLockRelease(ResGroupLock);
-
-		/* Add proc memory accounting info into group and slot */
-		groupIncMemUsage(group, slot, self->memUsage);
-
-		/* Update info in new slot */
-		pg_atomic_add_fetch_u32((pg_atomic_uint32*)&slot->nProcs, 1);
+		Assert(slot->sessionId = gp_session_id);
+		Assert(slot->memQuota = slotGetMemQuotaExpected(&caps));
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: s/=/==/

File: src/backend/utils/resgroup/resgroup.c [hunk 3]
```diff
@@ -1132,6 +1131,30 @@ groupDecMemUsage(ResGroupData *group, ResGroupSlotData *slot, int32 chunks)
...
+	groupDecMemUsage(group, slot, self->memUsage);
+	slot->nProcs--;
+}
+
+static void
+initSlot(ResGroupSlotData *slot, ResGroupCaps *caps)
+{
+	Assert(slot->inUse);
+
+	slot->sessionId = gp_session_id;
```
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: correct, i should set proc->mppSessionId instead for this case

File: src/backend/utils/resgroup/resgroup.c [hunk 4]
```diff
@@ -1132,6 +1131,30 @@ groupDecMemUsage(ResGroupData *group, ResGroupSlotData *slot, int32 chunks)
...
+{
+	groupIncMemUsage(group, slot, self->memUsage);
+	slot->nProcs++;
+}
+
+static void
+selfDetachSlot(ResGroupData *group, ResGroupSlotData *slot)
+{
+	groupDecMemUsage(group, slot, self->memUsage);
+	slot->nProcs--;
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: Is it possible to make a memory leak detection here? For example if `nProcs` is decreased to 0 then `slot->memUsage` is supposed to be 0?

File: src/backend/utils/resgroup/resgroup.c [hunk 5]
```diff
@@ -1994,22 +2008,39 @@ UnassignResGroupOnMaster(void)
...
+	else if (slot->nProcs == 0)
+	{
+		Assert(Gp_role == GP_ROLE_EXECUTE);
+
+		/* Release the slot memory */
+		groupReleaseMemQuota(group, slot);
+
+		/* Mark the slot as free */
+		selfUnsetSlot();
+		slot->inUse = false;
```
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: I don't agree.
we have `inUse == false` => `nProcs == 0`
but not always `nProcs == 0` => `inUse == false`

But i am sure we can remove the `inUse` in the later PR, which makes the slots a free list.

File: src/backend/utils/resgroup/resgroup.c (hunk 1 above)
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: yes

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: thanks

File: src/backend/utils/resgroup/resgroup.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Sure, then just keep `inUse` in this PR.

File: src/backend/utils/resgroup/resgroup.c (hunk 4 above)
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: yes

**Comment by pivotal-ning-yu** [comment]
LGTM

File: src/backend/utils/resgroup/resgroup.c [hunk 6]
```diff
@@ -1653,6 +1674,7 @@ wakeupSlots(ResGroupData *group)
 		Assert(waitProc->resWaiting != false);
 		Assert(waitProc->resSlotId == InvalidSlotId);
 
+		initSlot(&group->slots[slotId], &group->caps, waitProc->mppSessionId);
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: slotId should be checked here, when grant is false slotId would be -1, so the slot should not in inited.

Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Oops, please this comment, it's not for this PR.

**Pull request merged**